data/build_redial.py

```diff
@@ -0,0 +1,131 @@
+    write_jsonl("./redial/rd-test-formatted.jsonl", test_formatted)
+
+def length_summary(data_dir):
+    """prints a five number summary of the lengths of redial input/outputs."""
+    print("--Loading Dataset For Summary--")
+    train_data = read_jsonl(os.path.join(data_dir, "rd-train-formatted.jsonl"))
+    test_data = read_jsonl(os.path.join(data_dir, "rd-train-formatted.jsonl"))
+
+    def len_function(key):
+        return lambda x: len(x[key].split())
+
+    lengths = {"train_inputs" : list(map(len_function("conversation"), train_data)),
+                "train_targets" : list(map(len_function("response"), train_data)),
+                "test_inputs" : list(map(len_function("conversation"), test_data)),
+                "test_targets" : list(map(len_function("response"), test_data))}
+
+    for name, length_array in lengths.items():
+        print(name.upper() + ": ")
+        quartile_summary(length_array)
+
+#Helper Functions
+
+def read_jsonl(filename):
+    """Reads a jsonl file and returns as array of dicts."""
+    with open(filename, 'r') as json_file:
+        json_list = list(json_file)
+    data = []
+    for json_str in tqdm(json_list):
+        data.append(json.loads(json_str))
+    return data
+
+def write_jsonl(filename, arr):
+    """Writes array to jsonl."""
+    with open(filename, 'w') as f:
+        for line in arr:
+            f.write(json.dumps(line) + "\n")
+    
+def replace_ids(dialogue):
+    """Replaces movie ids in one redial dialogue with their corresponding movie titles."""
+    movie_titles = dialogue["movieMentions"]
+    for message in dialogue["messages"]:
+        text = message["text"]
+        replaced = []
+        for word in text.split():
+            if word[0] == "@" and re.sub('\\D', '', word) in movie_titles:
+                replaced.append( "@ " + movie_titles[re.sub('\\D', '', word)] + " @")
+            else:
+                replaced.append(word)
+        message["text"] = " ".join(replaced)
+
+def separate_responses(dataset):
+    """Creates a dataset of {"previous conversation" : "recommender response"} dictionaries
+    for every response by the recommending party in every conversation in the dataset."""
+    result = []
+    for dialogue in tqdm(dataset):
+        conversation = "" # the conversation history up until a turn
+        turn = "" # consecutive messags made by the same 
+        prev_id = None
+        for message in dialogue["messages"]:
+            if prev_id and message["senderWorkerId"] != prev_id:
+                if message["senderWorkerId"] == dialogue["initiatorWorkerId"]:
+                    # if the turn has switched to the user, add the (conversation, response) pair to response
+                    result.append({"conversation" : conversation.strip(), "response" : turn.strip()})
+                    conversation += " [Assistant] " + turn.strip()
+                else:
+                    conversation += " [User] " + turn.strip()
```

Comment:
This basically assumes that conversations are always started by the user, right? Mention that in a comment.

Reply:
It works either way. If the conversation is started by the assistant it adds {"conversation" : "", "response" : "first message"} to the data. I guess in that way there is always some sort of prompt to the assistant to generate a response, even if that prompt is empty. Is that what you meant?